Add implementation for MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS

stevew817 opened the pull request:
## Description
As was defined in #3878. This PR supersedes #3822. It uses the same building blocks, but with some changes:

* `mbedtls_psa_platform_get_builtin_key` is only responsible for the mapping of psa_key_id_t to lifetime and slot number. In #3822 it was also responsible for buffer allocation and key material copying, but that didn't make sense with the way `psa_crypto_slot_management.c` has evolved since then. The core can figure out the key size through the driver wrapper layer, and thus the platform code needn't be concerned with memory allocation.
* Reworked to fit into the current PSA Crypto core model.
* Added light test coverage by implementing the test driver's opaque `export_key` and `export_public_key` entry points.

There is a pending action to add support for testing opaque keys to the test driver, but that work is separate from enabling builtin key support. Therefore, actually testing the usage of builtin keys beyond through the export methods tested in this PR would be future work when opaque key testing gets tackled as a whole.

## Status
**READY**

## Requires Backporting
NO  (new feature)

## Migrations
NO (new feature)

## Todos
- [X] Tests
- [X] Documentation
- [X] Changelog updated


## Steps to test or reproduce
Test coverage added to `test_psa_crypto_drivers`.

tom-daubney-arm reviewed (approved):
Up to my level of understanding I can't see any issues with this PR. I have looked at the test coverage and this seems reasonable for now.

ronald-cron-arm commented on tests/src/drivers/key_management.c:
```diff
@@ -232,4 +336,63 @@ psa_status_t test_opaque_export_public_key(
+    psa_key_attributes_t *attributes,
+    uint8_t *key_buffer, size_t key_buffer_size, size_t *key_buffer_length )
+{
+#if defined(MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS)
```
Not necessary here thus I would remove.

ronald-cron-arm commented on tests/src/drivers/key_management.c:
```diff
@@ -223,6 +298,35 @@ psa_status_t test_opaque_export_public_key(
     uint8_t *data, size_t data_size, size_t *data_length )
 {
+#if defined(MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS)
+    if( psa_key_id_is_builtin( MBEDTLS_SVC_KEY_ID_GET_KEY_ID( psa_get_key_id( attributes ) ) ) )
```
The driver should rely only on the key slot number. The key identifier is and should be of no use to the driver.
If you remove this you can remove the guard above as well.

ronald-cron-arm commented on tests/src/drivers/key_management.c:
```diff
@@ -154,6 +178,57 @@ psa_status_t test_opaque_export_key(
     uint8_t *data, size_t data_size, size_t *data_length )
 {
+#if defined(MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS)
+    if( psa_key_id_is_builtin( MBEDTLS_SVC_KEY_ID_GET_KEY_ID( psa_get_key_id( attributes ) ) ) )
```
The driver should rely only on the key slot number. The key identifier is and should be of no use to the driver.
If you remove this you can remove the guard above as well.

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_driver_wrappers.function:
```diff
@@ -809,3 +809,107 @@ exit:
 }
 /* END_CASE */
+
+/* BEGIN_CASE depends_on:MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS */
```
We should probably add MBEDTLS_PSA_CRYPTO_DRIVERS as a dependency as well.

stevew817 replied:
Since the test is aimed specifically at the keys exposed through the test driver, I went one step further and added a dependency on both `MBEDTLS_PSA_CRYPTO_DRIVERS` and `PSA_CRYPTO_DRIVER_TEST`.

stevew817 commented on include/psa/crypto_extra.h:
```diff
@@ -711,6 +711,99 @@ psa_status_t mbedtls_psa_external_get_random(
+ *
+ * Values of this type are used to identify built-in keys.
+ */
+typedef uint64_t psa_drv_slot_number_t;
```
Unsure. This is what @gilles-peskine-arm set up in #3822...

gilles-peskine-arm replied:
The choice of type is meant to accommodate different ways of selecting slot numbers.

There are secure elements with a very small number of physical slots where dynamic allocation is not realistic: each slot is assigned a purpose at device integration time. This is usually the case on M-class devices with a secure element. On these devices, `uint8_t` would be plenty.

In larger applications, keys can be created in secure elements dynamically. Many secure elements do not have a way to choose a free “slot number” when creating a key. For example, if the secure element itself exposes a PSA crypto API interface, the ”slot number“ would naturally be the key id on the secure element, which the caller has to choose. With PKCS#11, the “slot number” would naturally be a label which the caller has to choose.

There are several potential ways to create a unique identifier when you don't know what identifiers have already been used. You can generate a random identifier, but it needs to be long enough in order for the probability of repeats to be negligible, much more than 64 bits. You can use a counter, which requires maintaining state, but guarantees uniqueness. A 32-bit counter can realistically wrap during the lifetime of a device, especially if a malicious actor (such as a breached application on a multi-application system) is allowed to create keys. A 64-bit counter will not wrap if your device's life span is at most a few decades.

Hence the choice of 64 bits, which is large enough to allow a simple way of choosing slot numbers while still being easy to manipulate (in particular, it's fixed-size).

> Having to handle uint64_t on embedded devices without CPU support does have a significant cost in terms of performance and code size.

How is this worse than `uint32_t[2]` or `uint8_t[8]`?

ronald-cron-arm replied:
If you use the uint64_t as an `uint32_t[2]` or `uint8_t[8]` that's not worse AFAIK. It is when you start to do 64 bits arithmetic on a 32 bits processor that there is a performance and code size cost compared to just 32 bits arithmetic. Not sure we expect much arithmetic to be done on slot numbers thus it could be all right.
Another possibility as the needs seem quite different depending on the platform would be for this type to be platform dependent. As long as the core does not have to do arithmetic and comparison with slot numbers this shouldn't be to painful to handle.

stevew817 replied:
Since there's currently no arithmetic being done on this type by the core (and comparison only happens inside of the test driver), are you okay with this PR using the proposed uint64_t for the reasons outlined by Gilles, @ronald-cron-arm? The type can always be turned into a platform specific type at a later point in time in case the little overhead turns out to be disruptive to a platform implementer.

ronald-cron-arm replied:
Yes, that's fine by me.

stevew817 commented on library/psa_crypto_driver_wrappers.c:
```diff
@@ -583,6 +593,34 @@ psa_status_t psa_driver_wrapper_export_public_key(
+    psa_key_attributes_t *attributes,
+    uint8_t *key_buffer, size_t key_buffer_size, size_t *key_buffer_length )
+{
+    psa_key_location_t location = PSA_KEY_LIFETIME_GET_LOCATION( attributes->core.lifetime );
```
Supposedly autogenerated file.

stevew817 commented on tests/src/helpers.c:
```diff
@@ -282,3 +282,38 @@ void mbedtls_param_failed( const char *failure_condition,
 }
 #endif /* MBEDTLS_CHECK_PARAMS */
+
+#if defined(MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS)
```
I'll move it into its own file. That'll make on-target testing for platform providers (like us) easier, since we'd just have to swap out the one file instead of modifying it to add more conditions on its definition.

stevew817 commented on include/psa/crypto_extra.h:
```diff
@@ -711,6 +711,99 @@ psa_status_t mbedtls_psa_external_get_random(
+ *                              an equivalent value, except that the key
+ *                              identifier field is set.
+ *                              On successful return, this function must set
+ *                              the attributes of the key: lifetime, type,
```
It's become lifetime instead of location, since I think the platform function should also be able to 'enforce' a driver-contained key to be treated as read-only while the driver would in theory be able to delete it.

stevew817 commented:
Had to rebase due to merge conflict.

I first pushed all of the changes requested by @ronald-cron-arm , and then did a clean rebase on top of development afterwards. The pre-rebase state is here: https://github.com/stevew817/mbedtls/tree/feature/driver_builtin_keys_1

yanesca commented:
There was a time-out on one of the tests, it is often sign of a CI glitch. I have restarted the jobs to see if this is the case.

yanesca commented:
The timeout went away, but apparently there is a linking issue as well on windows:
>81>mbedTLS.lib(psa_crypto_slot_management.obj) : error LNK2001: unresolved external symbol _mbedtls_psa_platform_get_builtin_key [C:\builds\workspace\mbed-tls-pr-head_PR-4160-head\visualc\VS2010\req_app.vcxproj]

ronald-cron-arm commented on tests/suites/test_suite_psa_crypto_driver_wrappers.function:
```diff
@@ -809,3 +809,107 @@ exit:
 }
 /* END_CASE */
+
+/* BEGIN_CASE depends_on:MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS */
```
Additional dependencies were added to `builtin_pubkey_export` but not here.

ronald-cron-arm commented on library/psa_crypto_driver_wrappers.c:
```diff
@@ -254,7 +254,17 @@ psa_status_t psa_driver_wrapper_get_key_buffer_size(
     {
 #if defined(PSA_CRYPTO_DRIVER_TEST)
-        case PSA_CRYPTO_TEST_DRIVER_LIFETIME:
+        case PSA_CRYPTO_TEST_DRIVER_LOCATION:
```
The documentation of this function should be updated.

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -276,6 +277,66 @@ static psa_status_t psa_load_persistent_key_into_slot( psa_key_slot_t *slot )
+    slot->attr = attributes.core;
+
+exit:
+    mbedtls_free( key_buffer );
```
If the function returns in error the clean-up could be done by `psa_wipe_key_slot()` in `psa_get_and_lock_key_slot()` but we have to take care not returning PSA_ERROR_DOES_NOT_EXIST when a key buffer has been allocated. Thus all together probably safer to free the key buffer in the key slot here.

stevew817 replied:
This comment is saying that you agree with the current construction, yes?

ronald-cron-arm replied:
The code has changed here thus this is anyway out-dated. But otherwise yes I was kind of thinking loud here.

ronald-cron-arm commented on include/psa/crypto_extra.h:
```diff
@@ -711,6 +711,99 @@ psa_status_t mbedtls_psa_external_get_random(
+            ( key_id <= MBEDTLS_PSA_KEY_ID_BUILTIN_MAX ) );
+}
+
+/** Platform function to obtain the location and slot of a built-in key.
```
```suggestion
/** Platform function to obtain the location and slot number of a built-in key.
```

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -293,25 +354,28 @@ psa_status_t psa_get_and_lock_key_slot( mbedtls_svc_key_id_t key,
     if( status != PSA_ERROR_DOES_NOT_EXIST )
         return( status );
 
-#if defined(MBEDTLS_PSA_CRYPTO_STORAGE_C)
```
I would keep here:
```C
#if defined(MBEDTLS_PSA_CRYPTO_STORAGE_C) ¦¦ defined(MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS)
```

stevew817 commented on include/psa/crypto_extra.h:
```diff
@@ -711,6 +711,99 @@ psa_status_t mbedtls_psa_external_get_random(
+ *
+ * \param key_id                The key ID for which to retrieve the
+ *                              location and slot attributes.
+ * \param[out] lifetime         On success, the lifetime associated with the key
```
I disagree. The platform should be able to indicate to the core whether a platform-installed key is read-only or modifiable.

ronald-cron-arm replied:
To be sure we are on the same page here, as I mentioned in my comment, in `psa-driver-interface.md` there is a specific point in the `acme_get_builtin_key` entry point specification that the driver may change the persistence level of the key from PSA_KEY_LIFETIME_PERSISTENT to something else:

```                                                                             
psa_status_t acme_get_builtin_key(psa_drv_slot_number_t slot_number,            
                                  psa_key_attributes_t *attributes,             
                                  uint8_t *key_buffer,                          
                                  size_t key_buffer_size,                       
                                  size_t *key_buffer_length);                   
```                                                                             
                                                                                
If this function returns `PSA_SUCCESS` or `PSA_ERROR_BUFFER_TOO_SMALL`, it must fill `attributes` with the attributes of the key (except for the key identifier). On success, this function must also fill `key_buffer` with the key context.
                                                                                
On entry, `psa_get_key_lifetime(attributes)` is the location at which the driver was declared and the persistence level `#PSA_KEY_LIFETIME_PERSISTENT`. The driver entry point may change the lifetime to one with the same location but a different persistence level. The standard attributes other than the key identifier and lifetime have the value conveyed by `PSA_KEY_ATTRIBUTES_INIT`.

Thus there is already a way for the core to be told if a key is read-only, through the driver that provides the description of the key. Is it the case that you want ALSO the platform function to be able to provide this information, for cases where the driver would not know the persistent level of the key (but all its other attributes ... ).

If we go for that, the specification of `acme_get_builtin_key` needs to be updated as the passed persistent level would not necessarily be `#PSA_KEY_LIFETIME_PERSISTENT`.

stevew817 replied:
If I recall correctly a discussion I had with @gilles-peskine-arm about this, there was a wish to support a use-case where a key in hardware might be erasable, but for the platform function (which maps the PSA IDs to a driver location + slot number) to be able to 'override' this, and not allow erase operations to be processed through the PSA Core. That'd mean the platform function would in the end be responsible for setting the PSA-facing lifetime value, and the driver-specific `get_builtin_key` hook would be allowed to 'override' the lifetime in case the platform tries to set it to something the driver doesn't support.

So IMO, the current flow adheres to this train of thought. I'll defer to @gilles-peskine-arm for more comments, and whether or not a doc update is needed.

gilles-peskine-arm replied:
Persistence levels are meant to help key management on devices that have a complex lifecycle. Typically it would indicate which keys must be erased on a factory reset, on a transfer of ownership, on a transfer of tenancy, or other similar events. We lump everything that isn't an application, our own generic library code or code that's closely associated with hardware (drivers) as “the platform”. The platform is where lifecycle-related decisions are made (at least after some initial manufacturing stages). So the platform gets to decide what the persistence level of a key is.

Does this mean that the platform should be able to override the persistence level reported by drivers? Arguably, no: the platform could decide the persistence level, and have the driver store this information and retrieve it. However, a built-in key may be stored in, say, a secure element that doesn't have room for this extra piece of metadata. In such a case, it would be onerous to require that the driver implement some additional metadata storage just to remember the persistence level (the driver could know the key type because that's an intrinsic property of the key, and it could even provide a sensible usage policy which the platform would only need to override if it's defending against a hostile application). So it makes sense to me that the platform could override the persistence level reported by the driver: the driver can just report the default level (or `READ_ONLY` for keys that are physically read-only), and the platform can override this for keys that must be preserved across device lifecycle stages.

I haven't looked at the code of this pull request in detail so I don't know exactly what this means in terms of implementation.

stevew817 replied:
The current state of this PR is that the full lifetime value for a key (location + persistence) is declared by the platform hook through mapping the builtin key ID. The lifetime value and key ID are then set onto a fresh key attributes structure, which is then passed onto the driver for completion, just as defined in `psa-driver-interface.md`.

The divergence here is that the current PR state opens up for the platform to declare a key with a different persistence than `PSA_KEY_LIFETIME_PERSISTENT`, which will be set to the key attributes struct before passing it onto `acme_get_builtin_key()`. This matches with the behaviour that you described in an inverse way - i.e. the platform attempts to override the persistency level on a key, and the driver entry point can complain if the platform-declared lifetime is fundamentally incompatible with the actual capabilities of the driver.

Is that acceptable, or does it actually need to happen in the order of driver reports default lifetime of key as known to the driver, and then the platform can override it to a platform-set value? With no validation from the driver-side to see whether the platform-set value is actually possible for the actual key?

I believe the current state allows for the lifetime override by the platform to happen in a sensible way. Do you agree, @gilles-peskine-arm ? In that case, I'll make one more commit to change the markdown documentation and make it clear that the key lifetime may be set to any value on entry to `get_builtin_key`, and the driver should report an error if the registered lifetime is fundamentally impossible for the requested key.

gilles-peskine-arm replied:
If there is no room to store the persistence level together with the key material, the two approaches are mostly equivalent: the only source for the persistent level is inside the platform (whether it's hard-coded or stored in some implementation-specific way). There's even a slight advantage to allowing the driver to validate the persistence level: it can prevent a platform from claiming that a key is erasable when the hardware knows that it isn't. I don't consider this a significant advantage because either way this is the sort of thing that would only be discovered by integration testing, and it's about as easy to test either way.

However, if there is room to store the persistence level with the key material, there is a definite advantage in allowing the driver to let the platform know what the persistence level is. This way, the persistence level can be decided at key provisioning time and stored together with the key's other metadata (type, policy, etc.).

How about keeping the code as it is, but changing the contract to allow the driver to override the lifetime in attributes? So if there's no room in the hardware, the driver just validates the persistence level (which I think only means to insist on `READ_ONLY` if the key is physically read-only). But if the hardware does store the persistence level, the driver may override the value in the attributes, and knowing this the platform can just put a default value and let the driver do its thing.

stevew817 replied:
> How about keeping the code as it is, but changing the contract to allow the driver to override the lifetime in attributes?

That's exactly what's implemented and what I proposed to do in my last comment. It sounds like you agree with that approach, so I'll update the doc to match the implementation ASAP.

ronald-cron-arm replied:
I've been through this. @gilles-peskine-arm thanks for the clarifications about what we mean by platform and its role in key lifetime management. I am happy with the proposed solution.

stevew817 replied:
I pushed the doc update in my last commit :)

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -293,27 +351,39 @@ psa_status_t psa_get_and_lock_key_slot( mbedtls_svc_key_id_t key,
-    return( PSA_ERROR_INVALID_HANDLE );
-#endif /* defined(MBEDTLS_PSA_CRYPTO_STORAGE_C) */
+#else /* MBEDTLS_PSA_CRYPTO_STORAGE_C || MBEDTLS_PSA_CRYPTO_BUILTIN_KEYS */
+    return( PSA_ERROR_DOES_NOT_EXIST );
```
I've been through the rebase and conflict resolution and this should be PSA_ERROR_INVALID_HANDLE.

stevew817 replied:
correct, good catch!

stevew817 replied:
I changed the offending commit with a forcepush, hope that's okay. I'll fix the other items in separate commits.

ronald-cron-arm replied:
That's okay, the force-pushed diff is clear about what has changed.

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -276,6 +277,66 @@ static psa_status_t psa_load_persistent_key_into_slot( psa_key_slot_t *slot )
+
+    /* If the key should exist according to the platform, load it through the
+     * driver interface. */
+    status = psa_driver_wrapper_get_key_buffer_size( &attributes,
```
This is not addressed and is still relevant I think. I can see it works with the test driver as the size of the buffer does not depend on the key type and size but that's only a specific case. A right sequence seems: call to `psa_driver_wrapper_get_builtin_key` with no buffer to get the key attributes, check for the returned error code to be PSA_ERROR_BUFFER_TOO_SMALL, call `psa_driver_wrapper_get_key_buffer_size`, allocate the buffer and then call again `psa_driver_wrapper_get_builtin_key`.

stevew817 replied:
Ah, okay. Now I see what you mean. Yes, let's do that sequence instead.

stevew817 replied:
Actually, wouldn't it make more sense to make the driver's `get_builtin_key` function return the expected length through the `key_buffer_length` output when it returns `BUFFER_TOO_SMALL`? That way seems cleaner to me, since there's not really a way to indicate through the key attributes alone that a key is built-in (i.e. the implementation of `get_key_buffer_size` is difficult to rely on when it cannot reliably distinguish whether you're asking for the buffer size of a builtin key or a 'regular' opaque key of the same algorithm, which might be two different sizes, as is the case in the test driver).

I'll make the relevant change to `psa-driver-interface.md` and then you and @gilles-peskine-arm can decide whether you agree.

ronald-cron-arm replied:
This has been proposed as part of the review of #3878 but decided no to go that way: see [here](https://github.com/ARMmbed/mbedtls/pull/3878#discussion_r561853623).

stevew817 replied:
reworked with your proposed flow :)

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -276,6 +277,66 @@ static psa_status_t psa_load_persistent_key_into_slot( psa_key_slot_t *slot )
+    status = psa_driver_wrapper_get_builtin_key(
+                slot_number, &attributes,
+                key_buffer, key_buffer_size, &key_buffer_length );
+    if( status != PSA_SUCCESS )
```
This does not seem to be addressed.

stevew817 replied:
It is now, thanks for pointing this out.

ronald-cron-arm commented on library/psa_crypto_slot_management.c:
```diff
@@ -276,6 +277,63 @@ static psa_status_t psa_load_persistent_key_into_slot( psa_key_slot_t *slot )
+
+exit:
+    if( status != PSA_SUCCESS )
+        psa_wipe_key_slot( slot );
```
I think this is a problem as you clear the key identifier and thus the load as a persistent key will always fail.

stevew817 replied:
You're right.

stevew817 replied:
Fixed now.

ronald-cron-arm commented:
The CI is passing now, just a CI glitch when running test_zlib_cmake.

ronald-cron-arm reviewed (changes requested):
Thanks for all your work on this. I only have three suggestions for small code improvements in psa_crypto_slot_management.c. Once we finalize https://github.com/ARMmbed/mbedtls/pull/4160#discussion_r598671102 this will be ok with me.

stevew817 commented on library/psa_crypto_slot_management.c:
```diff
@@ -276,6 +277,76 @@ static psa_status_t psa_load_persistent_key_into_slot( psa_key_slot_t *slot )
+    size_t key_buffer_size = 0;
+    size_t key_buffer_length = 0;
+
+    if( ! psa_key_id_is_builtin(
```
I believe this check logically fits into the sub-function, and the sub-function is static/local to this compilation unit, so it doesn't matter from an optimisation standpoint. A static function with only one reference will almost always be inlined automatically as soon as you enable any level of optimisation.

ronald-cron-arm replied:
ok fine by me but I am not a fan of "the compiler will deal with it" type of argument. Why make the compiler's work harder and not organize the C code as we would like it to be executed ?

stevew817 replied:
I wasn't solely arguing the compiler will deal with it, I was only arguing that 'saving a function call' isn't a heavyweight argument. My main point was that I think the check belongs in the subroutine, just like in the call tree beneath `psa_load_persistent_key_into_slot` the ID is checked to be a persistent ID (in effect by checking whether it exists on disk).

In addition, the reason we're doing an explicit check for the built-in range is to avoid the platform from being able to override keys outside of the built-in range. So it's a piece of logic that IMO definitely belongs in the subroutine which handles interaction with the platform hook.

ronald-cron-arm replied:
Thanks for those additional explanations. I understand better now what you meant by "this check logically fits into the sub-function" which I was not sure about the meaning in your first reply.

stevew817 commented on library/psa_crypto_slot_management.c:
```diff
@@ -276,6 +277,76 @@ static psa_status_t psa_load_persistent_key_into_slot( psa_key_slot_t *slot )
+    }
+
+    /* Check the platform function to see whether this key actually exists */
+    psa_set_key_id( &attributes, slot->attr.id );
```
Done

stevew817 commented on library/psa_crypto_slot_management.c:
```diff
@@ -276,6 +277,76 @@ static psa_status_t psa_load_persistent_key_into_slot( psa_key_slot_t *slot )
+        /* Builtin keys cannot be defined by the attributes alone */
+        if( status == PSA_SUCCESS )
+            status = PSA_ERROR_CORRUPTION_DETECTED;
+        goto exit;
```
Done

stevew817 commented:
Third rebase. The merge conflicts appeared due to #4257 adding to the bottom of the psa_crypto_driver_wrappers test suite, and this PR also contains additions to the bottom.

Resolved conflict by adding the new test cases from this PR after the test cases that were added in #4257. Previous state available at https://github.com/stevew817/mbedtls/tree/feature/driver_builtin_keys_3

ronald-cron-arm commented on docs/proposed/psa-driver-interface.md:
```diff
@@ -810,7 +810,7 @@ psa_status_t acme_get_builtin_key(psa_drv_slot_number_t slot_number,
 If this function returns `PSA_SUCCESS` or `PSA_ERROR_BUFFER_TOO_SMALL`, it must fill `attributes` with the attributes of the key (except for the key identifier). On success, this function must also fill `key_buffer` with the key context.
 
-On entry, `psa_get_key_lifetime(attributes)` is the location at which the driver was declared and the persistence level `#PSA_KEY_LIFETIME_PERSISTENT`. The driver entry point may change the lifetime to one with the same location but a different persistence level. The standard attributes other than the key identifier and lifetime have the value conveyed by `PSA_KEY_ATTRIBUTES_INIT`.
+On entry, `psa_get_key_lifetime(attributes)` is the location at which the driver was declared and a persistence level with which the platform is attempting to register the key. The driver entry point may choose to change the lifetime (`psa_set_key_lifetime(attributes, lifetime)`) of the reported key attributes to one with the same location but a different persistence level, in case the driver has ...
```
```suggestion
On entry, `psa_get_key_lifetime(attributes)` is the location at which the driver was declared and a persistence level with which the platform is attempting to register the key. The driver entry point may choose to change the lifetime (`psa_set_key_lifetime(attributes, lifetime)`) of the reported key attributes to one with the same location but a different persistence level, in case the driver has more specific knowledge about the actual persistence level of the key which is being retrieved. For example, if a driver knows it cannot delete a key, it may override the persistence level in the lifetime to `PSA_KEY_PERSISTENCE_READ_ONLY`. The standard attributes other than the key identifier and lifetime have the value conveyed by `PSA_KEY_ATTRIBUTES_INIT`.
```

stevew817 replied:
Done (during rebase)

ronald-cron-arm commented on include/psa/crypto_extra.h:
```diff
@@ -713,6 +713,103 @@ psa_status_t mbedtls_psa_external_get_random(
+ * \param[out] lifetime         On success, the lifetime associated with the key
+ *                              corresponding to \p key_id. Lifetime is a
+ *                              combination of which driver contains the key,
+ *                              and with what persistency the key is intended to
```
```suggestion
 *                              and with what persistence level the key is intended to
```

stevew817 replied:
Done (during rebase)

ronald-cron-arm commented on include/psa/crypto_extra.h:
```diff
@@ -713,6 +713,103 @@ psa_status_t mbedtls_psa_external_get_random(
+ *                              and with what persistency the key is intended to
+ *                              be used. If the platform implementation does not
+ *                              contain specific information about intended key
+ *                              persistence, the persistence may be reported as
```
```suggestion
 *                              persistence level, the persistence level may be reported as
```

stevew817 replied:
Done (during rebase)

ronald-cron-arm reviewed (changes requested):
Almost good to me, just some minor wording improvement suggestions. Otherwise in the "Rename test driver source files to avoid file name conflicts" commit you have to do the changes for AEAD test driver files as well now. After that the CI should be OK.

stevew817 commented:
Rebased the last couple of commits to address the review comments and adjust for the AEAD driver.

* 'Add test driver sources to VC build': added aead.c test driver source as well
* 'Rename test driver sources': renamed AEAD test driver source file as well
* 'Reword the builtin key language': applied language-related suggestions from @ronald-cron-arm

ronald-cron-arm reviewed (approved):
LGTM

ronald-cron-arm commented:
@tom-daubney-arm Please look and review the last version.

tom-daubney-arm reviewed (approved):
LGTM.

ronald-cron-arm commented:
@mpg @yanesca @gilles-peskine-arm could one of you have the final look and merge if this is ok.
